pkgs/top-level/all-packages.nix

```diff
@@ -16962,7 +16965,7 @@ with pkgs;
     };
   }).overrideAttrs (old: {
     configureFlags = old.configureFlags ++ ["--set=build.docs=false"];
-  });
+  }));
 
   makeRustPlatform = callPackage ../development/compilers/rust/make-rust-platform.nix { };
 
```

Comment:
To avoid confusion I think some of the argument names need to be changed from `rustc` to `rustc-unwrapped`... one of the lessons learned from `cc-wrapper` is that you want to deliberately make it hard for people to use the unwrapped compiler by accident, since it's almost always a mistake.  The diff below makes these changes:

```diff
diff --git a/pkgs/build-support/rust/rustc-wrapper/default.nix b/pkgs/build-support/rust/rustc-wrapper/default.nix
index 3f118a87b03f..d6034c08af47 100644
--- a/pkgs/build-support/rust/rustc-wrapper/default.nix
+++ b/pkgs/build-support/rust/rustc-wrapper/default.nix
@@ -1,30 +1,30 @@
-{ lib, runCommand, rustc, sysroot ? null }:
+{ lib, runCommand, rustc-unwrapped, sysroot ? null }:
 
-runCommand "${rustc.pname}-wrapper-${rustc.version}" {
+runCommand "${rustc-unwrapped.pname}-wrapper-${rustc-unwrapped.version}" {
   preferLocalBuild = true;
   strictDeps = true;
-  inherit (rustc) outputs;
+  inherit (rustc-unwrapped) outputs;
 
   env = {
-    prog = "${rustc}/bin/rustc";
+    prog = "${rustc-unwrapped}/bin/rustc";
     sysroot = lib.optionalString (sysroot != null) "--sysroot ${sysroot}";
   };
 
   passthru = {
-    inherit (rustc) pname version src llvm llvmPackages;
-    unwrapped = rustc;
+    inherit (rustc-unwrapped) pname version src llvm llvmPackages;
+    unwrapped = rustc-unwrapped;
   };
 
-  meta = rustc.meta // {
-    description = "${rustc.meta.description} (wrapper script)";
+  meta = rustc-unwrapped.meta // {
+    description = "${rustc-unwrapped.meta.description} (wrapper script)";
     priority = 10;
   };
 } ''
   mkdir -p $out/bin
-  ln -s ${rustc}/bin/* $out/bin
+  ln -s ${rustc-unwrapped}/bin/* $out/bin
   rm $out/bin/rustc
   substituteAll ${./rustc-wrapper.sh} $out/bin/rustc
   chmod +x $out/bin/rustc
-  ${lib.concatMapStrings (output: "ln -s ${rustc.${output}} \$${output}\n")
-    (lib.remove "out" rustc.outputs)}
+  ${lib.concatMapStrings (output: "ln -s ${rustc-unwrapped.${output}} \$${output}\n")
+    (lib.remove "out" rustc-unwrapped.outputs)}
 ''
diff --git a/pkgs/development/compilers/rust/default.nix b/pkgs/development/compilers/rust/default.nix
index 4738d25e0a17..d56b6fb9e08e 100644
--- a/pkgs/development/compilers/rust/default.nix
+++ b/pkgs/development/compilers/rust/default.nix
@@ -77,7 +77,7 @@ in
         inherit (bootstrapRustPackages) cargo rustc rustfmt;
       });
       rustc = wrapRustcWith {
-        rustc = self.rustc-unwrapped;
+        inherit (self) rustc-unwrapped;
         sysroot = if fastCross then self.rustc-unwrapped else null;
       };
       rustfmt = self.callPackage ./rustfmt.nix {
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 6321ba560180..c8700caec6cb 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -16931,8 +16931,8 @@ with pkgs;
     inherit (darwin) apple_sdk;
   };
 
-  wrapRustcWith = { rustc, ... } @ args: callPackage ../build-support/rust/rustc-wrapper args;
-  wrapRustc = rustc: wrapRustcWith { inherit rustc; };
+  wrapRustcWith = { rustc-unwrapped, ... } @ args: callPackage ../build-support/rust/rustc-wrapper args;
+  wrapRustc = rustc-unwrapped: wrapRustcWith { inherit rustc-unwrapped; };
 
   rust_1_73 = callPackage ../development/compilers/rust/1_73.nix {
     inherit (darwin.apple_sdk.frameworks) CoreFoundation Security SystemConfiguration;
@@ -16952,20 +16952,22 @@ with pkgs;
   inherit (rustPackages) cargo cargo-auditable cargo-auditable-cargo-wrapper clippy rustc rustPlatform;
 
   # https://github.com/NixOS/nixpkgs/issues/89426
-  rustc-wasm32 = wrapRustc ((rustc.unwrapped.override {
-    stdenv = stdenv.override {
-      targetPlatform = stdenv.targetPlatform // {
-        parsed = {
-          cpu.name = "wasm32";
-          vendor.name = "unknown";
-          kernel.name = "unknown";
-          abi.name = "unknown";
+  rustc-wasm32 = rustc.override {
+    rustc-unwrapped = (rustc.unwrapped.override {
+      stdenv = stdenv.override {
+        targetPlatform = stdenv.targetPlatform // {
+          parsed = {
+            cpu.name = "wasm32";
+            vendor.name = "unknown";
+            kernel.name = "unknown";
+            abi.name = "unknown";
+          };
         };
       };
-    };
-  }).overrideAttrs (old: {
-    configureFlags = old.configureFlags ++ ["--set=build.docs=false"];
-  }));
+    }).overrideAttrs (old: {
+      configureFlags = old.configureFlags ++ ["--set=build.docs=false"];
+    });
+  };
 
   makeRustPlatform = callPackage ../development/compilers/rust/make-rust-platform.nix { };
 
```

Reply:
Done. (Except I left rustc-wasm32 calling wrapRustc rather than doing an override, which I think is clearer.)